### PR TITLE
[WIP] Developer docs

### PR DESCRIPTION
**Pull request by nmearl**
Addresses #528.

### CODE REVIEW

File: docs/developer.rst [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+from some of the core aspects of the SpecViz architecture.
+
+.. image:: _static/specviz_diagram.png
+
+Data Model
+----------
+
+The central piece of SpecViz is the internal Qt data model expressed in the
+`DataListModel` class. It is responsible for maintaining the collection of
+:class:`~specutils.Spectrum1D` objects and exposing them as Qt `DataItem`s.
```
[the remark this answers is not in the document]
Comment (nmearl) [reply docs/developer.rst]: You don't think it'd be weird to talking about a `DataItem` without the context of a `DataItem` being used in a `DataListModel`?

File: docs/developer.rst [hunk 2]
```diff
@@ -0,0 +1,206 @@
+Developer Documentation
+=======================
+
+This set of documentation focuses on the structure of SpecViz, its various
+pieces, and how they work together. Below is a simplified reference diagram
+from some of the core aspects of the SpecViz architecture.
+
+.. image:: _static/specviz_diagram.png
```
[the remark this answers is not in the document]
Comment (nmearl) [reply docs/developer.rst]: Added another diagram. Let me know what you think!

File: docs/developer.rst [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+
+All spectral data are stored internally as :class:`specutils.Spectrum1D`, but
+these are exposed in SpecViz through |DataItem| objects. These objects are
+Qt-aware and can interact with the GUI in ways that a simple :class:`specutils.Spectrum1D`
+container cannot. However, individual |DataItem| objects are immutable in
+most cases, and only a single list of |DataItem| objects are ever maintained
+in the SpecViz application instance. Instead, |PlotDataItem| objects are used
+which internally reference a single |DataItem|. Multiple |PlotDataItem| objects
+can reference the same |DataItem|, this reduces duplication of data on the user's
+disk. Likewise, |PlotDataItem| objects have many mutable aspects that can be
```
Comment (astrofrog) [inline docs/developer.rst]: Minor: isn't it more a memory issue than a disk issue?

File: docs/developer.rst [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+reference to the |Workspace| object's |DataListModel|. It is the responsibility of the
+|PlotWindow| (and, more specifically, the |PlotWindow| object's |PlotWidget|) to
+create the corresponding |PlotProxyModel| used for that particular |PlotWindow|
+instance. In essence, the |PlotWindow| is really a container for housing the
+plot tool bar and the |PlotWidget|, and generally only contains functionality
+that doesn't directly involve manipulating the |PlotWidget| directly.
+
+The |PlotWidget| is the plotted representation of all the |PlotDataItem| objects in
+its internal |PlotProxyModel|. The widget itself is a subclass of `PyQtGraph` object's
+|PlotWidget| object. Anything that affects the visual representation of the
```
Comment (astrofrog) [inline docs/developer.rst]: duplicate 'object'

**Pull request merged**